File: 2025/scaffold.go
// go run scaffold.go

package main

import (
	"fmt"
	"os"
	"path/filepath"
)

func writeFile(path string, content string) {
	_ = os.WriteFile(path, []byte(content), 0644)
}

func main() {
	fmt.Println("Scaffolding Advent of Code 2025 in current directory...")
	rootDir, _ := os.Getwd()

	for day := 1; day <= 25; day++ {
		dirName := fmt.Sprintf("day_%02d", day)
		dirPath := filepath.Join(rootDir, dirName)
		_ = os.MkdirAll(dirPath, 0755)

		fmt.Printf("Creating files for %s...\n", dirName)

		// 1. input.txt (0 bytes)
		f, _ := os.Create(filepath.Join(dirPath, "input.txt"))
		f.Close()

		urlDay := fmt.Sprintf("%d", day)

		// 2. main.cpp (Template: 2 spaces)
		cppContent := fmt.Sprintf(`// https://adventofcode.com/2025/day/%[1]s
//
// clang++ -std=gnu++23 -stdlib=libc++ main.cpp -o main && ./main && rm main

#include <iostream>

int main() {
  std::cout << "Hello, Advent of Code 2025!" << std::endl;
}
`, urlDay)
		writeFile(filepath.Join(dirPath, "main.cpp"), cppContent)

		// 3. main.swift (Template: Top level)
		swiftContent := fmt.Sprintf(`// https://adventofcode.com/2025/day/%[1]s
//
// swiftc main.swift -o main && ./main && rm main
// -OR-
// swift main.swift

print("Hello, Advent of Code 2025!")
`, urlDay)
		writeFile(filepath.Join(dirPath, "main.swift"), swiftContent)

		// 4. main.go (Template: Hard Tab)
		// Raw string literal in Go preserves the Tab
		goContent := fmt.Sprintf(`// https://adventofcode.com/2025/day/%[1]s
//
// go build -o main main.go && ./main && rm main
// -OR-
// go run main.go

package main

import "fmt"

func main() {
	fmt.Println("Hello, Advent of Code 2025!")
}
`, urlDay)
		writeFile(filepath.Join(dirPath, "main.go"), goContent)

		// 5. main.rs (Template: 2 spaces)
		rsContent := fmt.Sprintf(`// https://adventofcode.com/2025/day/%[1]s
//
// rustc main.rs && ./main && rm main

fn main() {
  println!("Hello, Advent of Code 2025!");
}
`, urlDay)
		writeFile(filepath.Join(dirPath, "main.rs"), rsContent)

		// 6. main.zig (Template: 4 spaces)
		zigContent := fmt.Sprintf(`// https://adventofcode.com/2025/day/%[1]s
//
// zig build-exe main.zig && ./main && rm main
// -OR-
// zig run main.zig

const std = @import("std");

pub fn main() !void {
    std.debug.print("Hello, Advent of Code 2025!\n", .{});
}
`, urlDay)
		writeFile(filepath.Join(dirPath, "main.zig"), zigContent)
	}

	fmt.Println("Done! Happy coding!")
}

File: 2025/scaffold.rs
// rustc scaffold.rs && ./scaffold && rm scaffold

use std::fs;
use std::io::Write;

fn write_file(path: std::path::PathBuf, content: &str) -> std::io::Result<()> {
  let mut file = fs::File::create(path)?;
  file.write_all(content.as_bytes())?;
  Ok(())
}

fn main() -> std::io::Result<()> {
  println!("Scaffolding Advent of Code 2025 in current directory...");
  let root_dir = std::env::current_dir()?;

  for day in 1..=25 {
    let dir_name = format!("day_{:02}", day);
    let dir_path = root_dir.join(&dir_name);

    // Ignore error if exists
    let _ = fs::create_dir(&dir_path);
    println!("Creating files for {}...", dir_name);

    // 1. input.txt (0 bytes)
    fs::File::create(dir_path.join("input.txt"))?;

    let d = day.to_string();

    // 2. main.cpp (Template: 2 spaces)
    let cpp_content = format!(
r#"// https://adventofcode.com/2025/day/{d}
//
// clang++ -std=gnu++23 -stdlib=libc++ main.cpp -o main && ./main && rm main

#include <iostream>

int main() {{
  std::cout << "Hello, Advent of Code 2025!" << std::endl;
}}
"#);
    write_file(dir_path.join("main.cpp"), &cpp_content)?;

    // 3. main.swift (Template: Top level)
    let swift_content = format!(
r#"// https://adventofcode.com/2025/day/{d}
//
// swiftc main.swift -o main && ./main && rm main
// -OR-
// swift main.swift

print("Hello, Advent of Code 2025!")
"#);
    write_file(dir_path.join("main.swift"), &swift_content)?;

    // 4. main.go (Template: Hard Tab)
    // Raw string preserves the Tab character literally
    let go_content = format!(
r#"// https://adventofcode.com/2025/day/{d}
//
// go build -o main main.go && ./main && rm main
// -OR-
// go run main.go

package main

import "fmt"

func main() {{
	fmt.Println("Hello, Advent of Code 2025!")
}}
"#);
    write_file(dir_path.join("main.go"), &go_content)?;

    // 5. main.rs (Template: 2 spaces)
    let rs_content = format!(
r#"// https://adventofcode.com/2025/day/{d}
//
// rustc main.rs && ./main && rm main

fn main() {{
  println!("Hello, Advent of Code 2025!");
}}
"#);
    write_file(dir_path.join("main.rs"), &rs_content)?;

    // 6. main.zig (Template: 4 spaces)
    let zig_content = format!(
r#"// https://adventofcode.com/2025/day/{d}
//
// zig build-exe main.zig && ./main && rm main
// -OR-
// zig run main.zig

const std = @import("std");

pub fn main() !void {{
    std.debug.print("Hello, Advent of Code 2025!\n", .{{}});
}}
"#);
    write_file(dir_path.join("main.zig"), &zig_content)?;
  }

  println!("Done! Happy coding!");
  Ok(())
}
